File: app/src/main/java/com/caitlynwiley/pettracker/activities/CreateAccountActivity.java
package com.caitlynwiley.pettracker.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.caitlynwiley.pettracker.R;
import com.caitlynwiley.pettracker.models.Account;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import androidx.annotation.NonNull;

public class CreateAccountActivity extends BaseActivity {

    private EditText mEmail;
    private EditText mPassword;
    private EditText mPasswordRepeated;
    private TextView mErrorText;

    private FirebaseDatabase database = FirebaseDatabase.getInstance();
    private DatabaseReference ref = database.getReference();
    private FirebaseAuth mAuth;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.create_account);

        mAuth = FirebaseAuth.getInstance();
        mEmail = findViewById(R.id.email_field);
        mPassword = findViewById(R.id.password_field_one);
        mPasswordRepeated = findViewById(R.id.password_field_two);
        mErrorText = findViewById(R.id.error_msg);
        Button createBtn = findViewById(R.id.create_btn);
        createBtn.setOnClickListener(v -> {
            final String email = mEmail.getText().toString();
            // check passwords
            if (!mPassword.getText().toString().equals(mPasswordRepeated.getText().toString())) {
                hideKeyboard();
                showError("Passwords do not match. Please try again.");
                return;
            }

            Task<AuthResult> t = mAuth.createUserWithEmailAndPassword(email, mPassword.getText().toString());
            t.addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    // create account
                    Account a = new Account(mAuth.getUid(), email);
                    ref.child("users").child(mAuth.getUid()).setValue(a);
                    ref.child("users").child(mAuth.getCurrentUser().getUid()).child("num_pets").addValueEventListener(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                            if (dataSnapshot.getValue() == null || dataSnapshot.getValue().equals(0)) {
                                startActivity(new Intent(CreateAccountActivity.this, AddPetActivity.class));
                            } else {
                                startActivity(new Intent(CreateAccountActivity.this, MainActivity.class));
                            }
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError databaseError) {

                        }
                    });
                } else {
                    showError("Account creation failed with message: " + task.getException().getMessage());
                    hideKeyboard();
                }
            });
        });
    }

    private void showError(String errorMessage) {
        mErrorText.setText(errorMessage);
    }

    private void hideKeyboard() {
        InputMethodManager imm = (InputMethodManager) this.getSystemService(Activity.INPUT_METHOD_SERVICE);
        //Find the currently focused view, so we can grab the correct window token from it.
        View view = this.getCurrentFocus();
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = new View(this);
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }
}

File: app/src/main/java/com/caitlynwiley/pettracker/activities/CreateAccountActivity.kt
package com.caitlynwiley.pettracker.activities

import android.content.Intent
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.wrapContentWidth
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.preference.PreferenceManager
import com.caitlynwiley.pettracker.models.Account
import com.caitlynwiley.pettracker.theme.PetTrackerTheme
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class CreateAccountActivity: BaseActivity() {

    @ExperimentalAnimationApi
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val startBaseActivity = fun() {
            startActivity(Intent(this@CreateAccountActivity, BaseActivity::class.java))
        }

        setContent {
            Surface {
                PetTrackerTheme {
                    Scaffold(
                        topBar = {
                            TopAppBar(title = { Text("Create Account") })
                        }
                    ) {
                        Content(startBaseActivity)
                    }
                }
            }
        }
    }

    @Composable
    fun Content(startBaseActivity: () -> Unit) {
        val context = LocalContext.current
        val auth = FirebaseAuth.getInstance()
        var errorMsg by remember { mutableStateOf("") }
        val submit = fun(email: String, password: String, passwordAgain: String) {
            if (password != passwordAgain) {
                // hide keyboard
                errorMsg = "Passwords do not match. Please try again."
            } else {
                auth.createUserWithEmailAndPassword(email, password).addOnCompleteListener {
                    if (it.isSuccessful) {
                        val dbRef = FirebaseDatabase.getInstance().reference
                        val newAccount = Account(auth.uid!!, email)
                        dbRef.child("users").child(auth.uid!!).setValue(newAccount)
                        dbRef.child("users").child(auth.currentUser!!.uid).child("num_pets")
                            .addValueEventListener(object : ValueEventListener {
                                override fun onDataChange(dataSnapshot: DataSnapshot) {
                                    if (dataSnapshot.value == null || dataSnapshot.value == 0) {
                                        PreferenceManager.getDefaultSharedPreferences(context).edit()
                                            .putBoolean("creating_pet", true).apply()
                                        startBaseActivity()
                                    } else {
                                        PreferenceManager.getDefaultSharedPreferences(context).edit()
                                            .putBoolean("creating_pet", false).apply()
                                        startBaseActivity()
                                    }
                                }

                                override fun onCancelled(databaseError: DatabaseError) {}
                            })
                    } else {
                        errorMsg = it.exception!!.message.toString()
                    }
                }
            }
        }

        CreateAccount(onSubmit = submit, errorMsg)
    }

    @Composable
    fun CreateAccount(onSubmit: (String, String, String) -> Unit, errorMsg: String = "") {
        var email by remember { mutableStateOf("") }
        var password by remember { mutableStateOf("") }
        var passwordAgain by remember { mutableStateOf("") }

        Column(modifier = Modifier
            .fillMaxWidth()
            .padding(8.dp)) {
            Text("Email", modifier = Modifier.padding(4.dp))
            TextField(value = email, modifier = Modifier.fillMaxWidth(),
                onValueChange = { email = it })

            Text("Enter your password", modifier = Modifier.padding(4.dp))
            TextField(
                value = password,
                modifier = Modifier.fillMaxWidth(),
                onValueChange = { password = it },
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password)
            )

            Text("Enter your password again", modifier = Modifier.padding(4.dp))
            TextField(
                value = passwordAgain,
                modifier = Modifier.fillMaxWidth(),
                onValueChange = { passwordAgain = it },
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password)
            )

            Button(
                onClick = { onSubmit(email, password, passwordAgain) }, modifier = Modifier
                    .align(Alignment.CenterHorizontally)
                    .padding(top = 16.dp),
                colors = ButtonDefaults.buttonColors(backgroundColor = MaterialTheme.colors.secondary)
            )
            {
                Text("CREATE")
            }

            Text(
                errorMsg, color = MaterialTheme.colors.error,
                modifier = Modifier
                    .wrapContentWidth()
                    .align(Alignment.CenterHorizontally)
            )
        }
    }

    @Preview
    @Composable
    fun PreviewCreateAccount() {
        CreateAccount(fun(_, _, _) {})
    }
}